### PR TITLE
Final group capstone - Rent a house

### PR DESCRIPTION
**Pull request by yashodhicy**
### In this pull request our team created all Api requests according to requirements:

🌺 Add authentication
🌺 Add house api end points for create , view , show ,delete actions
🌺 Add reservation api end points for create , view actions
🌺 Install linter errors and fixed those

🌺api documentation link
https://house-rental-8mh7.onrender.com/api-docs/index.html

🌺 Front end app pull link
https://github.com/yashodhicy/Final-capstone-front-end/pull/28

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  # Include default devise modules.
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :trackable, :validatable, :omniauthable
+  include DeviseTokenAuth::Concerns::User
+  has_many :houses, dependent: :destroy
+  has_many :reservations, dependent: :destroy
+  validates :username, :email, presence: true
```
Comment (mcrd25) [inline app/models/user.rb:8]: [OPTIONAL] Consider a length validation for username

File: app/controllers/api/v1/houses_controller.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
+class Api::V1::HousesController < ApplicationController
+  before_action :authenticate_user!
+  before_action :set_houses, only: %i[index destroy]
+  def index
+    current_user
```
Comment (mcrd25) [inline app/controllers/api/v1/houses_controller.rb]: `current_user` is called but never user, please remove

File: app/controllers/api/v1/houses_controller.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    render json: @houses
+  end
+
+  def userhouses
+    user = current_user
+    @userhouses = user.houses.all
```
Comment (mcrd25) [inline app/controllers/api/v1/houses_controller.rb]: would it not be better simply call `@userhouses = current_user.houses.all`

File: app/controllers/api/v1/houses_controller.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    @house = user.houses.build(house_params)
+    if @house.save
+      render json: @house, status: :created
+    else
+      render json: { errors: @house.errors.full_messages }, status: :unprocessable_entity
+    end
+  end
+
+  def show
+    @house = House.find(params[:id])
```
Comment (mcrd25) [inline app/controllers/api/v1/houses_controller.rb:24]: Not seeing any error handling in the case where house cannot be found by given id

Comment (mcrd25) [reply app/controllers/api/v1/houses_controller.rb:24]: [OPTIONAL] This piece of code is also used in `destroy` maybe you can abstract it  into a method and make a `before_action` method

**Review by mcrd25 — CHANGES_REQUESTED** [review]
Hi Team,

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- Good routes 👍 
- Documentation present 👍 
- Good README 👍 
- Application works functionally 👍 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- [OPTIONAL] Should always have positive and negative tests to ensure it works the way you expect it. Eg. What do you expect the output to be when it's not valid or when valid etc.
Consider making all your tests more extensive, not just this one. For this simple application, it may not catch a possible crippling bug, but in bigger applications, it definitely can. Consider this for all projects moving forward.

**Hope all is well, and that my comments were helpful. 😸**

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
_Also, you do not have to implement all requested changes, as this is up to you in the end, but I strongly recommend you implement them._

File: app/models/user.rb (hunk 1 above)
Comment (yashodhicy) [reply app/models/user.rb:8]: Thank you

File: app/controllers/api/v1/houses_controller.rb (hunk 2 above)
Comment (yashodhicy) [reply app/controllers/api/v1/houses_controller.rb]: Thank you

**Review by Meltrust — APPROVED** [review]
_note: This review outcome only applies to the backend repo.  Kindly, go to the [frontend ](https://github.com/yashodhicy/Final-capstone-front-end/pull/28)repo to see the review for that project._

#### Hi @yashodhicy,


# **Wow, you did it 🎉**
Brilliant
## Thank you for the changes implemented 💪 🥇 ㊗️ 
#### _Unless you want to add more features, go ahead to your final presentation ⏩ ⏩ ⏩_ 
## You are about to finish the Microverse program.  You have come a long way!!!




### Good luck in the software industry!!  I'll see you there. ✨
### **Congratulations!!!!!!**  🎉
applause





Cheers and Happy coding!👏👏👏


------

**Comment by Mahamadou-Nouridine** [comment]
Thank you @Meltrust

**Pull request merged**